open-addressing HashTable implementation

Idolin opened the pull request.

Idolin committed:
open-addressing HashTable implementation

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
-    Object put(Object key, Object value) {
-        throw new UnsupportedOperationException();
+public class HashTable<K, V>
+        implements Map<K, V> {
```
Не было задания реализовывать какой-либо интерфейс

Idolin replied:
Мне так было привычнее

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+    private int size;
+    private int maxIndex;
+    private int threshold;
+    private double loadFactor;
```
Это переменная не должна меняться в течении жизни

Idolin replied:
Согласен
Добавлю final

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+        this.loadFactor = loadFactor;
+        if (capacity <= 0)
+            throw new IllegalArgumentException("Capacity must be greater then 0");
+        if (capacity > (1 << 30))
```
Зачем здесь этот сдвиг?

Idolin replied:
А как лучше написать 2**30?

Idolin replied:
Перепишу на Math.pow

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+    public HashTable(int capacity, double loadFactor) {
+        size = 0;
+        this.loadFactor = loadFactor;
+        if (capacity <= 0)
```
Форматирование неверное во всем методе

Idolin replied:
Исправлю

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+        if (loadFactor <= 0 || loadFactor > 1)
+            throw new IllegalArgumentException("loadFactor must be in range (0, 1]");
+        maxIndex = 4;
+        while (maxIndex < capacity) {
```
Зачем? И, если нужно, то нужно вынести в отдельный метод с говорящим названием

andrsuh replied:
Что-то типа добить до ближайшей степени двойки больше чем и тд, подумай

Idolin replied:
Хорошо, вынесу в отдельный метод

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+
+    @SuppressWarnings("unchecked")
+    private Entry<K, V>[] createArray(int length) {
+        return (Entry<K, V>[]) Entry.createArray(length);
```
В чем смысл выносить это в метод и делегировать статическому методу вложенного класса?

Idolin replied:
Вынес в отдельный метод, для удобства и чтобы пометить аннотацией @SuppressWarnings("unchecked")

Idolin replied:
Делегировал статическому методу вложенного класса, так как не знаю, как получить его тип извне

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+        return (Entry<K, V>[]) Entry.createArray(length);
+    }
+
+    private static class Entry<K, V> implements Map.Entry<K, V> {
```
Не просил реализовывать интерфейсы

Idolin replied:
Ориентировался на реализацию [HashTable](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/util/Hashtable.java) в jdk

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+        V value;
+        int overEnd;
+
+        protected Entry(K key, V value) {
```
А для чего `protected`?

Idolin replied:
Я ориентировался на реализацию [HashTable ](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/util/Hashtable.java)в jdk

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+    }
+
+    public HashTable(int capacity) {
+        this(capacity, 0.7f);
```
магическая константа

Idolin replied:
Я ориентировался на реализация [Hashtable ](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/util/Hashtable.java)в jdk
Там так же

Idolin replied:
Точнее там про неё написано в Javadoc комментарии

Idolin replied:
Добавлю комментарий

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+    }
+
+    public HashTable() {
+        this(65536);
```
магическая константа

Idolin replied:
Добавлю комментарий

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+    public V get(Object key) {
+        Entry<?, ?> entry = new Entry<>(key);
+        int position = getPosition(entry);
+        return (position >= 0) ? (table[position]).getValue() : null;
```
много ненужных скобок

Idolin replied:
Ну, я привык выделять такие выражения, которые возвращают boolean
Если вы считаете, что без скобок лучше смотрится, то могу убрать

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
-        throw new UnsupportedOperationException();
+    @Override
+    public boolean isEmpty() {
+        return (size == 0);
```
скобки зачем

Idolin replied:
Ну, я привык выделять такие выражения, которые возвращают boolean
Если вы считаете, что без скобок лучше смотрится, то могу убрать

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+    @Override
+    public boolean containsKey(Object key) {
+        int position = getPosition(new Entry<>(key));
+        return (position >= 0);
```
скобки

Idolin replied:
Ну, я привык выделять такие выражения, которые возвращают boolean
Если вы считаете, что без скобок лучше смотрится, то могу убрать

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
 
+    @Override
+    public boolean containsValue(Object value) {
+        for (int i = 0; i <= maxIndex; i++) {
```
Если уж решил реализовывать методы интерфеса Map, то реализовывал бы на О(1). А то со сдвигом так запарился, выиграл наносекунду, а тут по всей мапе ходит

Idolin replied:
Он и в [jdk реализации](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/util/Hashtable.java) тоже O(n)
Чтобы сделать его за O(1) нужно постоянно поддерживать актуальным ещё какой-то MultiSet, но это потери по памяти, производительности и никому не нужно

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+        }
+
+        public int getPosition(int maxIndex) {
+            return maxIndex & hash;
```
Мне нужны хорошие объяснения зачем все реализовано именно так

Idolin replied:
Для повышения производительности
Длина массива всегда равна 2^x + 1, где последний элемент не используется и всегда null
maxIndex же равен 2^x - 1, поэтому это наиболее быстрый способ вычислить позицию элемента по хэшу

Idolin replied:
Добавлю комментарии

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+        }
+
+        public int getSignedPosition(int maxIndex) {
+            return overEnd | getPosition(maxIndex);
```
аналогично

Idolin replied:
Скорость

Idolin replied:
Добавлю комментарии

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+        }
+
+        static protected Object createArray(int length) {
+            return Array.newInstance(Entry.class, length);
```
Зачем такой модный способ создания массива?

Idolin replied:
А как ещё создать массив generic-ов?

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+        }
+
+        boolean equals(Entry<?, ?> entry) {
+            return (hash == entry.hash) && key.equals(entry.key);
```
Откуда взята реализация метода?

Idolin replied:
Сам писал
А что тут не так?

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+        V value = table[position].getValue();
+        int shiftToPosition = position;
+
+        while (true) {
```
Слишком раздутый метод

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+            if (table[position] != null) {
+                if (entry.equals(table[position])) {
+                    --size;
+                    return (table[position]).setValue(entry.getValue());
```
зачем эти скобки везде?

Idolin replied:
Здесь случайно сохранились
Раньше здесь был cast
Исправлю

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+    private V putEntry(Entry<K, V> entry) {
+        int hashPosition = entry.getPosition(maxIndex);
+        int position = hashPosition;
+        while (true) {
```
я видел похожий код в remove, зачем нам дублирование?

Idolin replied:
Код конечно похожий, но не на столько, чтобы его можно было вынести в какой-либо отдельный метод
По крайней мере я никакого способа не вижу

Idolin replied:
Хотя...
Возможно
Надо подумать

Idolin replied:
Слегка подредактирую getPosition и использую его в putEntry вместо entry.getPosition
Это должно помочь убрать несколько строк из putEntry

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+        Entry<K, V>[] oldTable = table;
+        table = createArray(maxIndex + 1);
+        threshold = max((int) (loadFactor * maxIndex), 1);
+        --maxIndex;
```
довольно запутанно

Idolin replied:
Добавлю комментарии

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+            return overEnd | getPosition(maxIndex);
+        }
+
+        protected Entry<K, V> setSignPlus() {
```
protected?

Idolin replied:
Исправлю на public

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+            return this;
+        }
+
+        protected void setSignMinus() {
```
protected?

Idolin replied:
Исправлю на public

andrsuh commented on src/main/java/ru/itmo/java/HashTable.java:
```diff
@@ -1,23 +1,307 @@
+        }
+
+        protected void setSignMinus() {
+            overEnd = Integer.MIN_VALUE;
```
требуются объяснения

Idolin replied:
Добавлю комментарии

andrsuh reviewed (changes requested):
Для защиты можем договориться созвониться после правок. Как у нее производительность? Думаю, работает не слишком быстро?

Idolin reviewed (commented):
Работает на моём компьютере сравнимо с нативной реализацией Map
Может даже чуточку быстрее

Idolin committed:
Code is now much more clear
Added a lot of comments
Removed some "optimizations" (JIT optimizes some places really good)

Idolin commented:
Код поправил
По некотором вопросам написал комменты

Idolin committed:
In Entry (nested class): equals renamed to keyEquals
